ARROW-3769: [C++] Add support for reading non-dictionary encoded binary Parquet columns directly as DictionaryArray

hatemhelal opened the pull request:
This patch addresses the following JIRAS:

* [ARROW-3769](https://issues.apache.org/jira/browse/ARROW-3769): refactored record reader logic to toggle between the different builder depending on the column type (String or Binary) and the requested array type (Chunked "dense" or Dictionary).  These changes are covered by unittests and benchmarks.
* [PARQUET-1537](https://issues.apache.org/jira/browse/PARQUET-1537): fixed increment and covered by unittests.

Also included is an experimental class `ArrowReaderProperties` that can be used to select which columns are read directly as an `arrow::DictionaryArray`.  I think some more work is needed to fully address the requests in [ARROW-3772](https://issues.apache.org/jira/browse/ARROW-3772).  Namely, the ability automatically infer which columns in a parquet file should be read as `DictionaryArray`.  My current thinking is that this would be solved by introducing optional arrow type metadata to files written with the `parquet::arrow::FileWriter`.  There are some limitations with this approach but it would seem to satisfy the requests for users working with parquet files within the supported arrow ecosystem.

Note that the behavior with this patch is that incremental reading of a parquet file will not resolve the global dictionary for all of the row groups.  There are a few possible solutions for this:

* Introduce a concept of an "unknown" dictionary.  This will enable concatenating multiple row groups together so long as we define unknown dictionaries as equal (assuming indices have the same data type)
* Add an API for merging the schemas from multiple tables together.  This could be used after reading multiple row groups to enable concatenating the tables together into one.
* Add an API for inferring the global dictionary for the entire file.  This could be an expensive operation so ideally would be made optional.
* Allow a user-specified dictionary.  This could be useful in the limited case where a caller already knows the global dictionary list (computed through some other mechanism).

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+  state.SetBytesProcessed(state.iterations() * total_size_);
+}
+
+BENCHMARK_REGISTER_F(BM_PlainDecodingByteArray, DecodeArrow_Dict)->Range(1024, 65536);
```
Consider utilizing named constants for the Range values to increase readability here. Same comment for all other similar invocations of Range in this diff.

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+        ArrayFromJSON(::arrow::binary(), "[\"foo\", \"bar\", \"foo\", \"foo\"]");
+    num_values_ = static_cast<int>(expected_array_->length());
+
+    // arrow::BinaryType maps to parquet::ByteArray
```
This is very minor, and I may not be understanding this correctly, but should this comment say `arrow::BinaryArray` instead of `arrow::BinaryType`?

hatemhelal replied:
I can see that this is terse - the `expected_array_` is created with `arrow::DataType` equal to the `arrow::BinaryType`.  The loop is converting each value in the `expected_array_` into a `parquet::ByteArray`.

kevingurney replied:
Thanks for clarifying! That makes sense.

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+
+    // arrow::BinaryType maps to parquet::ByteArray
+    const auto& binary_array = static_cast<const ::arrow::BinaryArray&>(*expected_array_);
+    for (int64_t i = 0; i < binary_array.length(); i++) {
```
Very minor nitpick - I think you can use pre-increment (`++i`) instead of post-increment (`i++`) here and in other `for` loops in this diff in order to [avoid the temporary copy that is returned by the post-increment operator](https://en.cppreference.com/w/cpp/language/operator_incdec#Notes), which isn't required in this context. Doing a [quick search through the Arrow codebase](https://github.com/apache/arrow/search?l=C%2B%2B&q=for+%28int+i) seems to suggest that using pre-increment in similar scenarios is fairly common.

hatemhelal replied:
Good spot, I found this discussed in the [google c++ style guide](https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement).  I'll go for consistency with the rest of this codebase.

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
 
 BENCHMARK(BM_DictDecodingInt64_literals)->Range(1024, 65536);
 
+std::shared_ptr<::arrow::Array> MakeRandomStringsWithRepeats(size_t num_unique,
```
`MakeRandomStringsWithRepeats` seems like it could be useful elsewhere. Perhaps, this could be moved to somewhere under `arrow/cpp/src/arrow/util/`?

hatemhelal replied:
I must admit this isn't entirely original and based on a mashup of patterns I saw in other benchmarks and tests.  My thinking is that it would be good to somehow consolidate utilities like this.  Maybe `arrow/testing/random.h` is a good home for utilities like this?  I'll open a new JIRA issue to follow up on this as I think it would have wider downstream impact on the benchmarks and unittests.

hatemhelal replied:
Created [ARROW-4661](https://issues.apache.org/jira/browse/ARROW-4661) to revisit this.

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  void InitDataInputs() {
+    // Build a dictionary array and its dense representation for testing building both
+    auto dict_values = ::arrow::ArrayFromJSON(::arrow::binary(), "[\"foo\", \"bar\"]");
+    auto dtype = ::arrow::dictionary(::arrow::int8(), dict_values);
```
This is probably because I haven't worked with the `arrow::DictionaryType` before, and I didn't immediately realize that `::arrow::dictonary` is a factory function which returns an `arrow::DictionaryType` value, but I found the name `dtype` to be a bit overly generic in this context. I think it would be nice to use a consistent naming convention (e.g. `dict_type`) to parallel `dict_values`. Similarly, II think it would be nice to rename `indices` to `dict_indices`.

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
 
 BENCHMARK(BM_DictDecodingInt64_literals)->Range(1024, 65536);
 
+std::shared_ptr<::arrow::Array> MakeRandomStringsWithRepeats(size_t num_unique,
```
Consider adding a comment here which provides an example of what the output of `MakeRandomStringsWithRepeats` might look like for two example values of `num_unique` and `num_values`,

kevingurney replied:
Feel free to disregard this if you don't think it is helpful, but you could consider including `alphabet` (i.e. a list of letters from which to draw randomly) as one of the input arguments to `MakeRandomStringsWithRepeats`, rather than using `::arrow::random_ascii` inside of the function body. This might help facilitate greater flexibility/re-usability of this function in other contexts. For example, there might be a case in the future where a client might want to generate a random string containing Unicode characters, rather than just ASCII.

hatemhelal replied:
I like this idea and will take this up in [ARROW-4661](https://issues.apache.org/jira/browse/ARROW-4661)

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+  });
+
+  std::uniform_int_distribution<int64_t> indices_dist(0, num_unique - 1);
+  ::arrow::StringBuilder builder;
```
Since you know the number of values (`num_values`) that will be `Apppend`ed to the `arrow::StringBuilder`, it might be a good idea to call `builder.Reserve(num_values)` explicitly here.

hatemhelal replied:
Good catch!

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+
+BENCHMARK_DEFINE_F(BM_PlainDecodingByteArray, DecodeArrow_Dense)
+(benchmark::State& state) {
+  while (state.KeepRunning()) {
```
From the [google/benchmark `README.md`](https://github.com/google/benchmark#a-faster-keeprunning-loop): 

> "In C++11 mode, a ranged-based for loop should be used in preference to the KeepRunning loop for running the benchmarks."

Apparently, the generated assembly code for the range-based for loop in this context is slightly more efficient. However, looking at [other existing Parquet benchmarks](https://github.com/apache/arrow/blob/75c83578ced6de80015f3b93c1d290b9e74430e5/cpp/src/parquet/arrow/reader-writer-benchmark.cc#L210), the `while (state.KeepRunning())` paradigm seems to be widely in use in the Arrow project, so I will leave it up to other Arrow community members to comment further on whether the range-based for loop convention is worth following instead.

Personally, I think `while (state.KeepRunning())` is slightly more readable.

wesm replied:
We're starting to use the range-based loop in new benchmarks. I recommend using that

hatemhelal replied:
Thanks @wesm, I've updated the encoding benchmarks in the latest commit.

kevingurney commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -163,4 +172,132 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+    values_.reserve(num_values_);
+    total_size_ = 0;
+
+    for (int64_t i = 0; i < binary_array.length(); i++) {
```
According to my understanding, lines 212-229 are taking the random `arrow::Array` generated by `MakeRandomStringsWithRepeats` and "plain encoding" it as a Parquet `ByteArray`, with the intent that this `ByteArray` will be used as the test data for measuring the performance of decoding in the benchmark functions associated with this `benchmark::Fixture`.

Assuming my understanding is correct, consider factoring lines 212-229 into a helper function like `PlainEncodeByteArray(std::shared_ptr<::arrow::Array> array)`, or similar, to help make clear what the fixture `SetUp` process is doing at a high level. It would be nice to be able to glance at this `SetUp` code in order to understand the pre-conditions of the benchmark functions.

hatemhelal replied:
You understanding is correct and your suggestion makes this clearer.

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+    auto dtype = ::arrow::dictionary(::arrow::int8(), dict_values);
+    auto indices = ::arrow::ArrayFromJSON(::arrow::int8(), "[0, 1, 0, 0]");
+    ASSERT_OK(::arrow::DictionaryArray::FromArrays(dtype, indices, &expected_dict_));
+    expected_array_ =
```
Consider tweaking the name of this variable to something like `expected_dense_array` to make it easier to distinguish between the dictionary and dense representations of the test array.

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  void SetupEncoderDecoder() override {
+    encoder_ = MakeTypedEncoder<ByteArrayType>(Encoding::PLAIN);
+    decoder_ = MakeTypedDecoder<ByteArrayType>(Encoding::PLAIN);
+    encoder_->Put(input_data_.data(), num_values_);
```
Should this be wrapped in a call to `ASSERT_NO_THROW` like on line 436?

kevingurney commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  CheckDecodeDense(actual_num_values, builder);
+}
+
+TEST_F(TestDecodeArrowDict, DecodeDictionary) {
```
For consistency, consider either naming this `DecodeDict`, or spelling out `Dictionary` in full elsewhere in this diff.

kevingurney commented:
@hatemhelal I did an initial review of your changes, and overall they look good!

Almost all of my feedback is minor and primarily stylistic in nature. I am new to this area of the code base, so you can take my comments with a grain of salt. Many of them may stem more from my own lack of knowledge in this area, and my attempt to learn more, than from actual issues with the code.

Let me know if you have any questions regarding any of my feedback.

Thanks!

emkornfield commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +318,185 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+    ::arrow::ArrayVector actual_vec;
+    ASSERT_OK(builder.Finish(&actual_vec));
+    ASSERT_EQ(actual_vec.size(), 1);
+    ASSERT_TRUE(actual_vec[0]->Equals(expected_array_))
```
Can you use ASSERT_ARRAYS_EQUAL here?

hatemhelal replied:
Good tip!  I didn't know about `ASSERT_ARRAYS_EQUAL`.

hatemhelal commented:
Thanks @kevingurney and @emkornfield for the code review!  Let me know if you think of anything else.

rdmello commented on cpp/src/parquet/encoding.cc:
```diff
@@ -941,6 +948,12 @@ class DictByteArrayDecoder : public DictDecoderImpl<ByteArrayType>,
     return result;
   }
 
+  int DecodeArrowNonNull(int num_values, ::arrow::BinaryDictionaryBuilder* out) override {
```
I might be getting confused by the diff engine on GitHub, but is this the same code as on lines 725-730? If so, is there a common function both these methods could call?

hatemhelal replied:
They are the same but I don't see an easy way to share an implementation.  Let me think on this one.

hatemhelal replied:
I refactored this slightly in the latest commit.  @rdmello let me know how this looks to you.

hatemhelal commented:
The latest commit adds a test case for [PARQUET-1537](https://issues.apache.org/jira/browse/PARQUET-1537).  Just needed to have a null in the input data to reproduce the reported issue.

hatemhelal commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,141 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+  std::vector<std::string> dictionary(num_unique);
+
+  uint32_t seed = 0;
+  std::default_random_engine gen(seed);
```
Good point, I'll switch this to use `std::mt19937`

wesm commented:
We've largely stopped using MT in unit tests because it's slower than the alternatives (cc @pitrou)

pitrou commented:
What does MT mean in this context?

hatemhelal commented:
Would you recommend swapping back to the default random engine or another engine entirely?

pitrou commented:
The default random engine, or anything else that's fast (for example a xorshift-like PRNG). Are you worried about poor quality of the default engine?

hatemhelal commented:
I think default engine should be good reproducibility is the main concern for the benchmark.  Probably just an overreaction to "implementation defined"...

hatemhelal commented:
Switched back to `std::default_random_engine`.  I'd like to move this utility into the `testing/random.h` in [ARROW-4661](https://issues.apache.org/jira/browse/ARROW-4661)

wesm commented:
I can review this soon. Is this still WIP?

hatemhelal commented:
> I can review this soon. Is this still WIP?

I think this is ready to review.  This PR should resolve:

* [ARROW-3769](https://issues.apache.org/jira/browse/ARROW-3769)
* [PARQUET-1537](https://issues.apache.org/jira/browse/PARQUET-1537)

My plan is to use [ARROW-3772](https://issues.apache.org/jira/browse/ARROW-3772) to build on this change and plumb this through to the parquet -> arrow reader.

hatemhelal commented:
Most recent commit adds changes that go towards resolving [ARROW-3772](https://issues.apache.org/jira/browse/ARROW-3772)

* `ArrowReaderProperties` object that is used by `parquet::arrow::FileReader` to select whether or not to `use_threads` or read a particular column as a `DictionaryArray`.
* The `read_dictionary` flag is plumbed down through the column reading logic to toggle between either constructing a `DictionaryArray` or otherwise a chunked dense array.

@wesm can you let me know if this is heading in the right direction?  I'm working on some tests to accompany these changes.

hatemhelal commented:
Rebased on current arrow master.  Think this should resolve the CI failures...

hatemhelal commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -547,11 +558,19 @@ Status FileReader::Impl::ReadRowGroup(int row_group_index,
+    if (reader_properties_.read_dictionary(i)) {
+      auto dict_field =
+          std::make_shared<::arrow::Field>(schema->field(i)->name(), array->type());
+      RETURN_NOT_OK(schema->SetField(i, dict_field, &schema));
```
This fails to merge dictionaries across multiple row groups.  I think I can use `arrow::DictionaryType::Unify` to solve this.  Working on fixing this and adding tests.

wesm replied:
Might be better to "fix" the schema all at once rather than having to call `SetField` possibly multiple times. That can be done later...

wesm commented:
Will work on review for this today

wesm commented on cpp/src/parquet/arrow/arrow-reader-writer-test.cc:
```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
+    properties_ = default_arrow_reader_properties();
+  }
+
+  void InitFromJSON(const char* jsonStr) {
```
We use `snake_case` for variables

wesm commented on cpp/src/parquet/arrow/arrow-reader-writer-test.cc:
```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
+    //    Datum out;
+    //    FunctionContext ctx(default_memory_pool());
+    //    ASSERT_OK(DictionaryEncode(&ctx, Datum(dense_array), &out));
+    //    expected_dict_ = MakeSimpleTable(out.make_array(), /*nullable=*/true);
```
What is the error?

hatemhelal replied:
Swapping in the cast implementation for setting up the `expected_dict_` table results in an a test assertion failure in `AssertTablesEqual`.  I used a debugger to work out that the trouble is the dictionary generated by the cast has indices with type `arrow::Int32Type` while the builder generated one has indices with type `arrow::Int8Type`.

I don't know if this is actually an artifact of the trivial test cases that I used?

wesm commented on cpp/src/parquet/arrow/arrow-reader-writer-test.cc:
```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
+  CheckReadWholeFile(*expected_dense_);
+}
+
+const char* jsonInputs[] = {"[\"foo\", \"bar\", \"foo\", \"foo\"]",
```
camel case

wesm commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -682,6 +709,18 @@ Status OpenFile(const std::shared_ptr<::arrow::io::ReadableFileInterface>& file,
                   reader);
 }
 
+Status OpenFile(const std::shared_ptr<::arrow::io::ReadableFileInterface>& file,
```
This is deprecated I think, should be `::arrow::io::RandomAccessFile`. I opened https://issues.apache.org/jira/browse/ARROW-4789

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+static constexpr bool DEFAULT_USE_THREADS = false;
+static constexpr bool DEFAULT_READ_DICTIONARY = false;
+
+class PARQUET_EXPORT ArrowReaderProperties {
```
Hm, we might want to use a builder API for this, perhaps mark this experimental for now

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+
+  void set_read_dictionary(const int column_index, const bool read_dict) {
+    if (read_dict) {
+      read_dict_indices_.emplace(column_index);
```
Probably should use `insert` here since `column_index` is trivial

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+
+class PARQUET_EXPORT ArrowReaderProperties {
+ public:
+  explicit ArrowReaderProperties(const bool use_threads = DEFAULT_USE_THREADS,
```
We aren't really using `const` with primitive arguments

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+
+  bool use_threads() const { return use_threads_; }
+
+  void set_read_dictionary(const int column_index, const bool read_dict) {
```
remove const

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+      read_dict_indices_.erase(column_index);
+    }
+  }
+  bool read_dictionary(const int column_index) const {
```
remove const

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -51,6 +52,43 @@ class ColumnChunkReader;
+ public:
+  explicit ArrowReaderProperties(const bool use_threads = DEFAULT_USE_THREADS,
+                                 bool read_dict = DEFAULT_READ_DICTIONARY)
+      : use_threads_(use_threads), default_read_dict_(read_dict), read_dict_indices_() {}
```
If default_read_dict_ is globally configurable (does it need to be?) then add a setter for it

hatemhelal replied:
I think it didn't make a lot of sense to default read all columns as a dictionary array so I've removed the `default_read_dict_` logic entirely.  I want to flesh this out some more as part of native round-trip for `DictionaryArray` columns.

wesm commented on cpp/src/parquet/arrow/reader.h:
```diff
@@ -306,6 +345,12 @@ ::arrow::Status OpenFile(const std::shared_ptr<::arrow::io::ReadableFileInterfac
                          std::unique_ptr<FileReader>* reader);
 
+PARQUET_EXPORT
+::arrow::Status OpenFile(const std::shared_ptr<::arrow::io::ReadableFileInterface>& file,
```
API

wesm commented on cpp/src/parquet/arrow/record_reader.cc:
```diff
@@ -448,36 +448,17 @@ class RecordReader::RecordReaderImpl {
   void ResetDecoders() override { decoders_.clear(); }
 
-  inline void ReadValuesSpaced(int64_t values_with_nulls, int64_t null_count) {
+  virtual inline void ReadValuesSpaced(int64_t values_with_nulls, int64_t null_count) {
```
If this is virtual then `inline` probably should not be here

hatemhelal replied:
Thanks I wasn't too sure about that.  I found this [reference](http://www.cs.technion.ac.il/users/yechiel/c++-faq/inline-virtuals.html) which didn't help resolve my indecision.

wesm commented on cpp/src/parquet/arrow/record_reader.cc:
```diff
@@ -487,7 +468,7 @@ class TypedRecordReader : public RecordReader::RecordReaderImpl {
   }
 
-  inline void ReadValuesDense(int64_t values_to_read) {
+  virtual inline void ReadValuesDense(int64_t values_to_read) {
```
ditto

wesm commented on cpp/src/parquet/arrow/reader.cc:
```diff
@@ -1137,15 +1176,6 @@ struct TransferFunctor<
-        new_data->type = type;
-        chunks[i] = ::arrow::MakeArray(new_data);
-      }
-    }
```
Was this logic moved somewhere else, or is it simply not needed?

hatemhelal replied:
I believe this is no longer needed as the `StringDictionaryBuilder` overloads on the decoders take care of this directly.

wesm commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+
+    ::arrow::internal::ChunkedBinaryBuilder builder(
+        static_cast<int>(dict_buffer_->size()), ::arrow::default_memory_pool());
+    byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
```
This benchmark differs on only one line from the previous one, can you refactor to reuse code?

https://www.diffchecker.com/VPUwaQ0t

wesm commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+    auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder.get());
+
+    ::arrow::BinaryDictionaryBuilder builder(::arrow::default_memory_pool());
+    byte_array_decoder->DecodeArrow(num_values_, 0, valid_bits_, 0, &builder);
```
https://www.diffchecker.com/pPwQlQ9N

wesm commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+    byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
+  }
+
+  state.SetBytesProcessed(state.iterations() * total_size_);
```
https://www.diffchecker.com/rsvakCic

wesm commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+    decoder->DecodeArrowNonNull(num_values_, &builder);
+  }
+
+  state.SetBytesProcessed(state.iterations() * total_size_);
```
Differs from prior benchmark on one line

https://www.diffchecker.com/b9lpIkAM

wesm commented on cpp/src/parquet/encoding-benchmark.cc:
```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+    decoder->DecodeArrowNonNull(num_values_, &builder);
+  }
+
+  state.SetBytesProcessed(state.iterations() * total_size_);
```
https://www.diffchecker.com/0HrTFCY8

wesm commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+      } else {
+        ASSERT_OK(builder.Append(binary_array.GetView(i)));
+      }
+    }
```
Can you use `AppendArray` instead? https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_dict.cc#L243

hatemhelal replied:
Yes `AppendArray` is neater.

wesm commented on cpp/src/parquet/arrow/arrow-reader-writer-test.cc:
```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
+      } else {
+        ASSERT_OK(builder.Append(string_array.GetView(i)));
+      }
+    }
```
can you use AppendArray instead? 

https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_dict.cc#L243

wesm commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+    std::shared_ptr<::arrow::Array> actual;
+    ASSERT_OK(builder.Finish(&actual));
+    ASSERT_ARRAYS_EQUAL(*actual, *expected_str_dict_);
+  }
```
Can you address this code duplication?

wesm commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  ::arrow::StringDictionaryBuilder builder(default_memory_pool());
+  auto actual_num_values = decoder_->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
```
Seems like the code duplication in these 4 test functions ought to be addressable

wesm commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+  ASSERT_NE(byte_array_decoder, nullptr);
+  auto actual_num_values = byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
+  CheckDecodeDict(actual_num_values, builder);
+}
```
More code duplication, these 5 months are all "shaped" very similarly

wesm commented on cpp/src/parquet/encoding-test.cc:
```diff
@@ -314,6 +325,258 @@ TEST(TestDictionaryEncoding, CannotDictDecodeBoolean) {
+}
+
+const char* json[] = {"[\"foo\", \"bar\", \"foo\", \"foo\"]",
+                      "[\"foo\", \"bar\", \"foo\", null]"};
```
Can you generate general richer random examples? These ones aren't very interesting

wesm commented on cpp/src/parquet/encoding.cc:
```diff
@@ -748,11 +769,11 @@ class PlainByteArrayDecoder : public PlainDecoder<ByteArrayType>,
         ARROW_RETURN_NOT_OK(out->AppendNull());
       }
       bit_reader.Next();
+      ++i;
```
Does this fix the bug that had been reported elsewhere? Is this fix covered by a new test added here?

hatemhelal replied:
Yes, and yes, from an earlier [commit](https://github.com/apache/arrow/pull/3721#issuecomment-467493421):

> The latest commit adds a test case for PARQUET-1537. Just needed to have a null in the input data to reproduce the reported issue.

wesm commented on cpp/src/parquet/encoding.cc:
```diff
@@ -934,13 +956,34 @@ class DictByteArrayDecoder : public DictDecoderImpl<ByteArrayType>,
+    int result = 0;
+    PARQUET_THROW_NOT_OK(DecodeArrowNonNull(num_values, out, &result));
+    return result;
+  }
```
The feeling of code duplication is indeed suspicious. A templated impl class could help

hatemhelal replied:
I reworked the `ByteArrayDecoder` interface to:

* Use a public interface templated over the builder types
* Introduced a `WrappedBuilderInterface` as a sort of shim layer between the real builder instance and the one used within the decoder
* Added private virtual interface for decoding that works on a `WrappedBuilderInterface`

Not sure about the names but let me know if you have any more thoughts on this implementation.

wesm commented on cpp/src/parquet/arrow/arrow-reader-writer-test.cc:
```diff
@@ -2420,6 +2420,81 @@ TEST(TestArrowWriterAdHoc, SchemaMismatch) {
+}
+
+const char* jsonInputs[] = {"[\"foo\", \"bar\", \"foo\", \"foo\"]",
+                            "[\"foo\", \"bar\", \"foo\", \"null\"]"};
```
Consider generating higher quality test data? I think that this param-based testing might be a bit more elaborate than needed

hatemhelal replied:
I wanted to make sure the tests covered both the `DecodeArrow` and `DecodeArrowNonNull` code paths.  There might be other ways of achieving this but parameterization seemed convenient to me.

hatemhelal replied:
> Consider generating higher quality test data?

Completely agree: I'll try to refactor the data generation from the benchmarks to share it with these test cases.

hatemhelal commented on cpp/src/parquet/encoding.h:
```diff
@@ -35,6 +35,7 @@
 namespace arrow {
 
 class BinaryDictionaryBuilder;
+class StringDictionaryBuilder;
```
These forward declarations can be removed

hatemhelal commented:
This patch addresses the following JIRAS:

* [ARROW-3769](https://issues.apache.org/jira/browse/ARROW-3769): refactored record reader logic to toggle between the different builder depending on the column type (String or Binary) and the requested array type (Chunked "dense" or Dictionary).  These changes are covered by unittests and benchmarks.
* [PARQUET-1537](https://issues.apache.org/jira/browse/PARQUET-1537): fixed increment and covered by unittests.

Also included is an experimental class `ArrowReaderProperties` that can be used to select which columns are read directly as an `arrow::DictionaryArray`.  I think some more work is needed to fully address the requests in [ARROW-3772](https://issues.apache.org/jira/browse/ARROW-3772).  Namely, the ability automatically infer which columns in a parquet file should be read as `DictionaryArray`.  My current thinking is that this would be solved by introducing optional arrow type metadata to files written with the `parquet::arrow::FileWriter`.  There are some limitations with this approach but it would seem to satisfy the requests for users working with parquet files within the supported arrow ecosystem.

Note that the behavior with this patch is that incremental reading of a parquet file will not resolve the global dictionary for all of the row groups.  There are a few possible solutions for this:

* Introduce a concept of an "unknown" dictionary.  This will enable concatenating multiple row groups together so long as we define unknown dictionaries as equal (assuming indices have the same data type)
* Add an API for merging the schemas from multiple tables together.  This could be used after reading multiple row groups to enable concatenating the tables together into one.
* Add an API for inferring the global dictionary for the entire file.  This could be an expensive operation so ideally would be made optional.
* Allow a user-specified dictionary.  This could be useful in the limited case where a caller already knows the global dictionary list (computed through some other mechanism).

@wesm, let me know if you have any further feedback on this patch.  My preference would be to merge this (+/- naming, trivialities, obvious mistakes) and keep working on the open design issues summarized here as follow up JIRAs/PRs.  Just let me know how you'd like to proceed.

wesm commented:
Ah well the build is passed so merging now

xhochy commented:
Huge thanks for tackling this @hatemhelal !
